lib/gc/Transforms/Utils/VectorUtils.cpp

```diff
@@ -0,0 +1,480 @@
+            .Case<vector::TransferWriteOp>(
+                [&](vector::TransferWriteOp writeOp) {
+                  return moveBack(op, operationPosition);
+                })
+            .Case<vector::BroadcastOp>([&](vector::BroadcastOp bcOp) {
+              return moveFront(op, operationPosition);
+            })
+            .Default([&](Operation *op) { return success(); });
+    if (failed(ret)) {
+      LDBG("Wrong to move operation:" << *op << "\n");
+      return;
+    }
+  }
+}
+
+// Need to move some operations like extract_slice or insert_slice.
+// Because those operation may interpret our analysis result. e.g.:
+// ```
+// clang-format off
+  // %21 = vector.transfer_read %18[%c0, %c0], %cst {in_bounds = [true, true]} : tensor<16x16xf32>, vector<16x16xf32>
+  // %22 = arith.addf %21, %20 : vector<16x16xf32>
+  // %23 = vector.transfer_write %22, %extracted_slice_12[%c0, %c0] {in_bounds = [true, true]} : vector<16x16xf32>, tensor<16x16xf32>
+  // %inserted_slice_13 = tensor.insert_slice %18 into %arg14[%arg13, 0] [16, 16] [1, 1] : tensor<16x16xf32> into tensor<32x16xf32>
+  // %extracted_slice_14 = tensor.extract_slice %arg16[%arg13, 0] [16, 16] [1, 1] : tensor<32x16xf32> to tensor<16x16xf32>
+  // %24 = vector.transfer_read %cst_0[%c0, %c0], %cst {in_bounds = [true, true]} : tensor<16x16xf32>, vector<16x16xf32>
+  // %25 = arith.maximumf %22, %24 : vector<16x16xf32>
+  // %26 = vector.transfer_write %25, %extracted_slice_14[%c0, %c0] {in_bounds = [true, true]} : vector<16x16xf32>, tensor<16x16xf32>
+  // %inserted_slice_15 = tensor.insert_slice %23 into %arg15[%arg13, 0] [16, 16] [1, 1] : tensor<16x16xf32> into tensor<32x16xf32>
+  // %inserted_slice_16 = tensor.insert_slice %26 into %arg16[%arg13, 0] [16, 16] [1, 1] : tensor<16x16xf32> into tensor<32x16xf32>
+// clang-format on
+// ```
+// The maximumf and addf operation can be a same group, but the extract_slice
+// operation interpret us.
+// The move operation(extra_slice) will check its parameters. In order to
+// ensure that it does not affect the correctness of the result, we will only
+// move the moved op after the op to which the parameters belong to. If it's
+// operand is all the block argument, we will move it to the begining of the
+// block.
+// insert_slice just move them to the privious of the first operation which
+// use it.
+void moveOpsFrontOrBack(func::FuncOp *func, MLIRContext *ctx,
+                        std::function<bool(Operation *)> &conditionalFunc) {
+  // Pre-order traversal of each op
+  // Record each operation position. Inorder to we can kown current operation
+  // should move after which operation.
+  DenseMap<Operation *, size_t> operationPosition;
+  SmallVector<Operation *, 8> candidateOps;
+  size_t opCounter = 0;
+
+  // get the position of each operation
+  func->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    operationPosition[op] = opCounter++;
+    if (conditionalFunc(op))
+      candidateOps.emplace_back(op);
+  });
```

Comment:
This traverse looks expensive and redundant if  `moveOpsFrontOrBack` would be frequently called while the `func` is just same.